adr/2025_03_10-suggestions/README.md

```diff
@@ -0,0 +1,178 @@
+
+If I were to share my wish-list for features this should have, it would be able to answer questions like:
+
+- Given a `clientID`, what `userID` made this change?
+- At what time was this `clientID` first used?
+  - This gives approximate times for attributing content to a time, without having to store an additional timestamp per change).
+  - Most people don't need the specific time, they just want to know relative time "Is this days, months or years old?"
+  - Auditing should be done at a different level, and not in a user writable data structure like a YDoc
+
+I'd like to have Kevin give more details on the sort of features/properties we can expect from this. It will be a crucial feature to the implementation, and likely to be worked on first since it will be foundational to how the versioning works.
+
+## Summary
+
+This document was purposefully written from the perspective that Y.js is one of many document storage and synchronization mechanisms. And, we would like to keep our editor de-coupled from these specific implementation details as much as possible to encourage code re-use between different implementations.
+
+## Appendix
+
+This is a collection of scattered thoughts on the _diff_-erent aspects of diffing  (sorry, had to), and their complications.
+
+### Keeping user intent, while diffing
+
+Diffing nested data structures (e.g. HTML) is fundamentally different than diffing by the semantics of a document. This, is probably best described in an example:
+
+Take the following document:
+
+```html
+<p>Some text</p>
+<p>More text</p>
+```
+
+Let's say that the user deletes the range of text from "text" and collapses the two paragraphs into 1 paragraph, resulting in this document:
+
+```html
+<p>Some More text<p>
+```
+
+If we examine the changes, this is actually multiple operations:
+
+- deletion of the word "text"
+- insertion of the words "More text" at the position where "text" was
+- deletion of 2nd paragraph
+
+If we were to represent this change using HTML elements it was meant to look like this:
+
+```html
+<p>Some <del>text<p>
+<p></del>More text</p>
+```
+
+Yet, a diff would represent it like this:
+
+```html
+<p>Some <del>text</del><ins>More text</ins></p>
+<del><p>More text</p></del>
+```
+
+They ultimately result in the same in the final document but, the difference is in their semantics and user expectations. There are probably ways to simplify these changes in such a way that it is closer to what the user actually intended with the change (e.g. Prosemirror's [simplifyChanges](https://github.com/ProseMirror/prosemirror-changeset)), but it gets complicated rather quickly and we have to face that it may not always be easy to represent/display.
+
+Whatever the approach, we need to make sure that the changes are displayed in a way that is easy to understand and aligns with the user's intent.
+
```

Comment:
Make clear: this is a tradeoff of a nested element approach, but we are purposefully not choosing to implement a flat YText.

Reply:
Consider other limitations of the diff-based approach.

 - Possibly, undo-redo may change the result of the diff
 - moving content shows large deletes & insertions